cleangoldenami/main.go

```diff
@@ -58,23 +60,60 @@ func main() {
 				log.Fatalf("error fetching images for region %v: %v", regionName, err)
 			}
 
-			if imageCount := len(images); imageCount < quota {
+			if imageCount := len(images); (quota - imageCount) >= desiredImageCapacity {
 				if *verbose {
 					fmt.Printf("Region %v is under quota. (Images: %v, Quota: %v)\n", regionName, imageCount, quota)
 				}
 			} else {
-				imageToDelete, err := getOldestImage(images)
-				if err != nil {
-					log.Fatalf("error determining which image to delete in region %v: %v", regionName, err)
-				}
-				if *dryRun {
-					fmt.Printf("Region %v is at quota (%v) - would delete %v\n", regionName, quota, *imageToDelete.ImageId)
-				} else {
-					err := deregisterImage(ec2Client, imageToDelete)
+				numOfImagesToDelete := desiredImageCapacity - (quota - imageCount)
+
+				arm64Images := filterImages(images, "arm64", "rhel-arm64")
+				legacyx86Images := filterImages(images, "x86_64", "legacy-x86_64")
+				x86Images := filterImages(images, "x86_64", "rhel-x86_64")
+
+				var imagesToDelete []ec2Types.Image
+
+				for !imageDeletionCheck(imagesToDelete, numOfImagesToDelete, arm64Images, legacyx86Images, x86Images) {
+					// Returns the most populated image type and the slice containing images of that type
+					mostPopulatedImagesByType, mostPopulatedImageType := getMostPopulatedImageType(arm64Images, legacyx86Images, x86Images)
+
+					// Returns the oldest image from the most populated type's slice along with its index value
+					ImageToDeleteIndex, ImageToDelete, err := getOldestImage(mostPopulatedImagesByType)
 					if err != nil {
-						log.Fatalf("error deregistering image %v in region %v: %v", *imageToDelete.ImageId, regionName, err)
+						// Log a fatal error if there was an issue retrieving the oldest image
+						log.Fatalf("error determining which image to delete in region %v: %v", regionName, err)
+					}
+
+					// If ImageToDeleteIndex is valid (non-negative), proceed with deletion
+					if ImageToDeleteIndex >= 0 {
+						// Add the image to the list of images to delete
+						imagesToDelete = append(imagesToDelete, ImageToDelete)
+
+						// Remove the deleted image from its respective slice based on its type
+						switch mostPopulatedImageType {
+						case "arm64Images":
+							arm64Images = append(arm64Images[:ImageToDeleteIndex], arm64Images[ImageToDeleteIndex+1:]...)
+						case "legacyx86Images":
+							legacyx86Images = append(legacyx86Images[:ImageToDeleteIndex], legacyx86Images[ImageToDeleteIndex+1:]...)
+						case "x86Images":
+							x86Images = append(x86Images[:ImageToDeleteIndex], x86Images[ImageToDeleteIndex+1:]...)
+						}
+					}
+				}
+				for _, image := range imagesToDelete {
+					for i, t := range image.Tags {
+						if *t.Key == "version" {
```

Comment:
Do we need this if-statement? At this point we know all of the images we need to delete, I think we should be able to run right through it.

Reply:
the if statement is there so that we can output the version tag when we deregister the AMI. I initially tried to have it output like *image.Tags[0].Value, but there no guarantee that the version tag will always be the first entry.

Reply:
Ah, I actually really like this a lot - good call! +1 to keeping it. As a nano-optimization it may be worth adding a `break` as the last thing in this if-statement so we don't needlessly check other tags once we've found `version`.